// sanity-plugin-datatable/build/table.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _lodash = require('lodash');

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _primitives = require('./primitives');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Table = function Table(props) {
  var data = props.data,
      rows = props.rows,
      columns = props.columns,
      handleChange = props.handleChange;

  /**
  * Render rows
  */

  var rowsList = (0, _lodash.range)(rows).map(function (rowIndex) {
    return _react2.default.createElement(
      _primitives.Row,
      { key: rowIndex },
      (0, _lodash.range)(columns).map(function (columnIndex) {
        return _react2.default.createElement(_primitives.Cell, {
          key: columnIndex,
          type: 'text',
          value: (data[rowIndex] || [])[columnIndex] || '',
          onChange: function onChange(e) {
            return handleChange(e, rowIndex, columnIndex);
          }
        });
      })
    );
  });

  /**
   * Render
   */

  return _react2.default.createElement(
    _primitives.Container,
    null,
    rows === 0 && _react2.default.createElement(_primitives.Empty, { children: 'Empty table' }),
    rowsList
  );
};

Table.propTypes = {
  data: _propTypes2.default.array,
  rows: _propTypes2.default.number,
  columns: _propTypes2.default.number,
  handleChange: _propTypes2.default.func
};

exports.default = Table;
